File: src/main/java/org/lanternpowered/porygen/points/PointsGenerator.java
/*
 * This file is part of Porygen, licensed under the MIT License (MIT).
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) contributors
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the Software), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED AS IS, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package org.lanternpowered.porygen.points;

import com.flowpowered.math.vector.Vector2d;
import org.lanternpowered.porygen.GeneratorContext;
import org.lanternpowered.porygen.map.CellGenerator;
import org.lanternpowered.porygen.map.Cell;
import org.lanternpowered.porygen.util.Rectangled;

import java.util.List;
import java.util.Random;

public interface PointsGenerator {

    /**
     * Populates the {@link List} with point coordinates that will be
     * used by the {@link CellGenerator} to generate {@link Cell}s.
     *
     * @param context The world that will be generated
     * @param random The random that should be used
     * @param rectangle The rectangle that represents the area
     * @return The result points
     */
    List<Vector2d> generatePoints(GeneratorContext context, Random random, Rectangled rectangle);
}

File: src/main/kotlin/org/lanternpowered/porygen/points/PointsGenerator.kt
/*
 * Porygen
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.porygen.points

import org.lanternpowered.porygen.GeneratorContext
import org.lanternpowered.porygen.map.Cell
import org.lanternpowered.porygen.math.geom.Rectangled
import org.lanternpowered.porygen.util.random.Xor128Random
import org.spongepowered.math.vector.Vector2d
import kotlin.random.Random

/**
 * A generator that will generate points within a [Rectangled].
 */
interface PointsGenerator {

  /**
   * Generates a [Collection] with point coordinates that will be
   * used to generate [Cell]s.
   *
   * @param context The context that contains extra information related to the generation
   * @param rectangle The rectangle that represents the area that the points should be located in
   * @param random The random used to generate the points
   * @return A collection with the generated points
   */
  fun generatePoints(
      context: GeneratorContext, rectangle: Rectangled, random: Random = Xor128Random(context.seed)
  ): Collection<Vector2d>
}
